File: phantom-mac-app/Sources/Phantom/Services/LiveCopilotOrchestrator.swift
import Foundation

@MainActor
final class LiveCopilotOrchestrator {
    typealias ModelStream = (@escaping (String) -> Void, @escaping () -> Void) async throws -> String

    func execute(
        allowedEntityIds: [String],
        allowedDocumentIds: [String],
        questionText: String,
        retrievalAvailable: Bool,
        hasActiveEvidence: Bool,
        preferredDocumentsForDecision: @escaping (LiveTurnDecision) -> [String] = { _ in [] },
        firstModel: @escaping ModelStream,
        retrieve: @escaping (LiveTurnDecision) async throws -> LiveCopilotRetrieval,
        secondModel: @escaping (LiveTurnDecision, LiveCopilotRetrieval) -> ModelStream,
        publish: @escaping (String) -> Void,
        resetPublishedAttempt: @escaping () -> Void,
        protocolRejected: ((String) -> Void)? = nil,
        decisionParsed: ((LiveTurnDecision, Int, Bool) -> Void)? = nil
    ) async throws -> LiveCopilotResult {
        _ = retrievalAvailable
        _ = hasActiveEvidence
        var modelCalls = 0
        var protocolRetries = 0
        var parser = PhantomControlFrameParser(allowedEntityIds: allowedEntityIds, allowedDocumentIds: allowedDocumentIds)
        var firstResponse = ""
        var fallbackBuffer = ""

        while true {
            modelCalls += 1
            do {
                var streamProtocolError: PhantomProtocolError?
                firstResponse = try await firstModel({ chunk in
                    guard streamProtocolError == nil else { return }
                    do {
                        let visible = try parser.feed(chunk)
                        if !visible.isEmpty { publish(visible) }
                    } catch let error as PhantomProtocolError {
                        streamProtocolError = error
                    } catch {
                        streamProtocolError = PhantomProtocolError(code: "control_json_invalid")
                    }
                }, {
                    parser = PhantomControlFrameParser(allowedEntityIds: allowedEntityIds, allowedDocumentIds: allowedDocumentIds)
                    resetPublishedAttempt()
                })
                if let streamProtocolError { throw streamProtocolError }
                if !parser.hasReceivedChunks, !firstResponse.isEmpty {
                    let visible = try parser.feed(firstResponse)
                    if !visible.isEmpty { publish(visible) }
                }
                _ = try parser.complete()
                break
            } catch let error as PhantomProtocolError {
                var candidate = parser.fallbackAnswerText()
                if candidate.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    candidate = PhantomControlFrameParser.extractBareAnswer(firstResponse)
                }
                if !candidate.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    fallbackBuffer = candidate
                }
                protocolRejected?(error.code)
                if Self.isCompleteAnswer(fallbackBuffer) {
                    protocolRetries = max(protocolRetries, 1)
                    protocolRejected?("control_frame_fallback")
                    let fallbackDecision = PhantomControlFrameParser.fallbackAnswerDecision()
                    decisionParsed?(fallbackDecision, modelCalls, false)
                    publish(fallbackBuffer)
                    return LiveCopilotResult(
                        answer: fallbackBuffer, decision: fallbackDecision, modelCallCount: modelCalls,
                        protocolRetryCount: protocolRetries, retrievalStatus: "not_requested", activeEvidence: []
                    )
                }
                if protocolRetries == 0 {
                    protocolRetries += 1
                    parser = PhantomControlFrameParser(allowedEntityIds: allowedEntityIds, allowedDocumentIds: allowedDocumentIds)
                    resetPublishedAttempt()
                    continue
                }
                throw error
            }
        }

        var decision = try parser.complete()
        if decision.action == .retrieve, decision.retrievalQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            decision = LiveTurnDecision(
                action: decision.action,
                questionType: decision.questionType,
                intent: decision.intent,
                answerBasis: decision.answerBasis,
                entityType: decision.entityType,
                entityId: decision.entityId,
                retrievalQuery: LiveCopilotRetrievePolicy.normalizeRetrievalQuery(decision: decision, questionText: questionText),
                preferredDocumentIds: decision.preferredDocumentIds,
                targetSeconds: decision.targetSeconds,
                allowCode: decision.allowCode,
                confidence: decision.confidence
            )
        }
        decisionParsed?(decision, modelCalls, false)
        if protocolRetries > 0, decision.action == .retrieve {
            throw PhantomProtocolError(code: "control_repair_retrieve_invalid")
        }
        if decision.action != .retrieve {
            let answer = PhantomControlFrameParser.resolveAnswer(firstResponse)
            guard Self.isCompleteAnswer(answer) else {
                throw BackendError.server("The AI provider returned an incomplete response.")
            }
            return LiveCopilotResult(
                answer: answer, decision: decision, modelCallCount: modelCalls,
                protocolRetryCount: protocolRetries, retrievalStatus: "not_requested", activeEvidence: []
            )
        }

        let retrieval = try await retrieve(decision)
        var finalResponse = ""
        modelCalls += 1
        let completed = try await secondModel(decision, retrieval)({ chunk in
            finalResponse += chunk
            publish(chunk)
        }, {
            finalResponse = ""
            resetPublishedAttempt()
        })
        if finalResponse.isEmpty { finalResponse = completed }
        guard Self.isCompleteAnswer(finalResponse) else {
            throw BackendError.server("The AI provider returned an incomplete response.")
        }
        return LiveCopilotResult(
            answer: finalResponse, decision: decision, modelCallCount: modelCalls,
            protocolRetryCount: protocolRetries, retrievalStatus: retrieval.status, activeEvidence: retrieval.snippets
        )
    }

    static func extractBody(_ response: String) -> String {
        PhantomControlFrameParser.extractAnswerBody(response)
    }

    nonisolated static func isCompleteAnswer(_ response: String) -> Bool {
        guard !response.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        return response.components(separatedBy: "```").count % 2 == 1
    }
}

File: phantom-windows-app/Services/LiveCopilotOrchestrator.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using SecureOverlay.Domain;
using SecureOverlay.Domain.Entities;

namespace SecureOverlay.Services
{
    public sealed class LiveCopilotOrchestrator
    {
        public delegate Task<(string Response, string Error)> ModelStream(
            Action<string> onChunk,
            Action onRetryCleanup,
            CancellationToken cancellationToken);

        public async Task<LiveCopilotResult> ExecuteAsync(
            IEnumerable<string> allowedEntityIds,
            IEnumerable<string> allowedDocumentIds,
            string questionText,
            bool retrievalAvailable,
            bool hasActiveEvidence,
            ModelStream firstModel,
            Func<LiveTurnDecision, CancellationToken, Task<LiveCopilotRetrieval>> retrieve,
            Func<LiveTurnDecision, LiveCopilotRetrieval, ModelStream> secondModel,
            Action<string> publish,
            Action? resetPublishedAttempt,
            CancellationToken cancellationToken,
            Func<LiveTurnDecision, IReadOnlyList<string>>? preferredDocumentsForDecision = null,
            Action<string>? protocolRejected = null,
            Action<LiveTurnDecision, int, bool>? decisionParsed = null)
        {
            _ = retrievalAvailable;
            _ = hasActiveEvidence;
            var modelCalls = 0;
            var protocolRetries = 0;
            PhantomControlFrameParser parser;
            var firstResponse = string.Empty;
            var fallbackBuffer = string.Empty;

            while (true)
            {
                parser = new PhantomControlFrameParser(allowedEntityIds, allowedDocumentIds);
                modelCalls++;
                try
                {
                    var first = await firstModel(
                        chunk =>
                        {
                            var visible = parser.Feed(chunk);
                            if (visible.Length > 0) publish(visible);
                        },
                        () =>
                        {
                            parser = new PhantomControlFrameParser(allowedEntityIds, allowedDocumentIds);
                            resetPublishedAttempt?.Invoke();
                        },
                        cancellationToken).ConfigureAwait(false);
                    firstResponse = first.Response;
                    if (!string.IsNullOrEmpty(first.Error)) throw new InvalidOperationException(first.Error);
                    if (!parser.HasReceivedChunks && first.Response.Length > 0)
                    {
                        var visible = parser.Feed(first.Response);
                        if (visible.Length > 0) publish(visible);
                    }
                    parser.Complete();
                    break;
                }
                catch (PhantomProtocolException error)
                {
                    var candidate = parser.GetFallbackAnswerText();
                    if (string.IsNullOrWhiteSpace(candidate))
                    {
                        candidate = PhantomControlFrameParser.ExtractBareAnswer(firstResponse);
                    }

                    if (!string.IsNullOrWhiteSpace(candidate))
                    {
                        fallbackBuffer = candidate;
                    }

                    protocolRejected?.Invoke(error.Code);
                    if (IsCompleteAnswer(fallbackBuffer))
                    {
                        protocolRetries = Math.Max(protocolRetries, 1);
                        protocolRejected?.Invoke("control_frame_fallback");
                        var fallbackDecision = PhantomControlFrameParser.FallbackAnswerDecision();
                        decisionParsed?.Invoke(fallbackDecision, modelCalls, false);
                        publish(fallbackBuffer);
                        return new LiveCopilotResult(
                            fallbackBuffer, fallbackDecision, modelCalls, protocolRetries, "not_requested",
                            Array.Empty<RetrievedContextSnippet>());
                    }

                    if (protocolRetries == 0)
                    {
                        protocolRetries++;
                        resetPublishedAttempt?.Invoke();
                        continue;
                    }

                    throw;
                }
            }

            var decision = parser.Decision!;
            if (decision.Action == LiveCopilotAction.Retrieve)
            {
                var query = string.IsNullOrWhiteSpace(decision.RetrievalQuery)
                    ? LiveCopilotRetrievePolicy.NormalizeRetrievalQuery(decision, questionText)
                    : decision.RetrievalQuery;
                var docs = decision.PreferredDocumentIds.Count > 0
                    ? decision.PreferredDocumentIds
                    : preferredDocumentsForDecision?.Invoke(decision) ?? Array.Empty<string>();
                decision = decision with { RetrievalQuery = query, PreferredDocumentIds = docs };
            }

            decisionParsed?.Invoke(decision, modelCalls, false);
            if (protocolRetries > 0 && decision.Action == LiveCopilotAction.Retrieve)
                throw new PhantomProtocolException("control_repair_retrieve_invalid");
            if (decision.Action != LiveCopilotAction.Retrieve)
            {
                var answer = PhantomControlFrameParser.ResolveAnswer(firstResponse);
                if (!IsCompleteAnswer(answer)) throw new InvalidOperationException("The AI provider returned an incomplete response.");
                return new LiveCopilotResult(
                    answer, decision, modelCalls, protocolRetries, "not_requested", Array.Empty<RetrievedContextSnippet>());
            }

            var retrieval = await retrieve(decision, cancellationToken).ConfigureAwait(false);
            var finalResponse = string.Empty;
            var second = secondModel(decision, retrieval);
            modelCalls++;
            var completed = await second(
                chunk =>
                {
                    finalResponse += chunk;
                    publish(chunk);
                },
                () =>
                {
                    finalResponse = string.Empty;
                    resetPublishedAttempt?.Invoke();
                },
                cancellationToken).ConfigureAwait(false);
            if (!string.IsNullOrEmpty(completed.Error)) throw new InvalidOperationException(completed.Error);
            if (finalResponse.Length == 0) finalResponse = completed.Response;
            if (!IsCompleteAnswer(finalResponse)) throw new InvalidOperationException("The AI provider returned an incomplete response.");
            return new LiveCopilotResult(finalResponse, decision, modelCalls, protocolRetries, retrieval.Status, retrieval.Snippets);
        }

        public static bool IsCompleteAnswer(string response)
        {
            if (string.IsNullOrWhiteSpace(response)) return false;
            return response.Split("```", StringSplitOptions.None).Length % 2 == 1;
        }

        public static string ExtractBody(string response)
            => PhantomControlFrameParser.ExtractAnswerBody(response);
    }
}
